src/server/qtquick/wrenderhelper.cpp

```diff
@@ -566,6 +572,108 @@ QWRenderer *WRenderHelper::createRenderer(QWBackend *backend)
     return renderer;
 }
 
+constexpr const char *GraphicsApiName(QSGRendererInterface::GraphicsApi api)
+{
+    switch (api) {
+        using enum QSGRendererInterface::GraphicsApi;
+    case Software:
+        return "Software";
+    case OpenGL:
+        return "OpenGL";
+    case Vulkan:
+        return "Vulkan";
+    default:
+        return "Unknown/Unsupported";
+    }
+}
+
+void WRenderHelper::setupRendererBackend(QWBackend *testBackend)
+{
+    const auto wlrRenderer = qgetenv("WLR_RENDERER");
+
+    if (wlrRenderer == "auto" || wlrRenderer.isEmpty()) {
+        if (qEnvironmentVariableIsSet("QSG_RHI_BACKEND")) {
+            // when environment variable QSG_RHI_BACKEND was set, don't call setGraphicsApi
+            return;
+        }
+
+        QList<QSGRendererInterface::GraphicsApi> apiList = {
+            QSGRendererInterface::OpenGL,
+            QSGRendererInterface::Software
+        };
+        QWDisplay *display = nullptr;
+        if (!testBackend) {
+            display = new QWDisplay();
+            testBackend = QWBackend::autoCreate(display);
+
+            if (!testBackend)
+                qFatal("Failed to create wlr_backend");
+        }
+        QQuickWindow::setGraphicsApi(WRenderHelper::probe(testBackend, apiList));
+        if (display)
+            wl_display_destroy(display->handle());
+    } else if (wlrRenderer == "gles2") {
+        QQuickWindow::setGraphicsApi(QSGRendererInterface::OpenGL);
+    } else if (wlrRenderer == "vulkan") {
+#ifdef ENABLE_VULKAN_RENDER
+        QQuickWindow::setGraphicsApi(QSGRendererInterface::Vulkan);
+#else
+        qFatal("Vulkan support is not enabled");
+#endif
+    } else if (wlrRenderer == "pixman") {
+        QQuickWindow::setGraphicsApi(QSGRendererInterface::Software);
+    } else {
+        qFatal() << "Unknown/Unsupported wlr renderer: " << wlrRenderer;
+    }
+}
+
+QSGRendererInterface::GraphicsApi WRenderHelper::probe(QWBackend *testBackend, const QList<QSGRendererInterface::GraphicsApi> &apiList)
+{
+    auto acceptApi = QSGRendererInterface::Unknown;
+
+    for (auto api : apiList) {
+        auto renderer = createRenderer(testBackend, api);
+        if (!renderer) {
+            qInfo() << GraphicsApiName(api) << " api failed to create wlr_renderer";
+            continue;
+        }
+
+        const auto *formats = renderer->getDmabufTextureFormats();
+
+        if (formats && formats->len == 0) {
+            qInfo() << GraphicsApiName(api) << " api don't support any format";
+            continue;
+        }
+
+        // TODO: how to test when formats gets NULL
+        if (formats && formats->len) {
+            auto *format = &formats->formats[0];
+            auto *alloc = wlr_allocator_autocreate(testBackend->handle(), renderer->handle()); // destroy follow backend
```

Comment:
Where is the codes of `destroy follow backend`

Reply:
backend_destroy -> finish_drm_renderer -> wlr_allocator_destroy

Reply:
I can't reproduce the crash，you are right， there are not double free